libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/incremental/ClasspathSnapshot.kt

```diff
@@ -54,31 +54,30 @@ open class ClasspathSnapshot protected constructor(
             return KaptClasspathChanges.Unknown
         }
 
-        loadEntriesFor(changedFiles)
+        val unchangedBetweenCompilations = dataForFiles.keys.intersect(previousSnapshot.dataForFiles.keys).filter { it !in changedFiles }
+        val currentToLoad = dataForFiles.keys.filter { it !in unchangedBetweenCompilations }.also { loadEntriesFor(it) }
+        val previousToLoad = previousSnapshot.dataForFiles.keys.filter { it !in unchangedBetweenCompilations }
 
-        val currentHashAbiSize = changedFiles.sumBy { dataForFiles[it]!!.classAbiHash.size }
-        val currentHashesToAnalyze =
-            HashMap<String, ByteArray>(currentHashAbiSize).also { hashes ->
-                changedFiles.forEach {
-                    hashes.putAll(dataForFiles[it]!!.classAbiHash)
-                }
-            }
-
-        val currentUnchanged = dataForFiles.keys.filter { it !in changedFiles }
-        val previousChanged = previousSnapshot.dataForFiles.keys.filter { it !in currentUnchanged }
-
-        check(changedFiles.size == previousChanged.size) {
+        check(currentToLoad.size == previousToLoad.size) {
             """
-            Number of changed files in snapshots differs. Reported changed files: $changedFiles
+            Number of loaded files in snapshots differs. Reported changed files: $changedFiles
             Current snapshot data files: ${dataForFiles.keys}
             Previous snapshot data files: ${previousSnapshot.dataForFiles.keys}
         """.trimIndent()
         }
 
-        val previousHashAbiSize = previousChanged.sumBy { previousSnapshot.dataForFiles.get(it)?.classAbiHash?.size ?: 0 }
+        val currentHashAbiSize = currentToLoad.sumBy { dataForFiles[it]!!.classAbiHash.size }
+        val currentHashesToAnalyze =
+            HashMap<String, ByteArray>(currentHashAbiSize).also { hashes ->
+                currentToLoad.forEach {
+                    hashes.putAll(dataForFiles[it]!!.classAbiHash)
+                }
+            }
+
+        val previousHashAbiSize = previousToLoad.sumBy { previousSnapshot.dataForFiles.get(it)?.classAbiHash?.size ?: 0 }
```

Comment:
Seems impossible that `previousSnapshot` has nullable entries for `previousToLoad`.
If this is the case, then creating a map for hashes could be extracted:
```
private fun hashesFor(filesToLoad: Collection<File>): Map<String, ByteArray> {
        val size = filesToLoad.sumBy { dataForFiles[it]!!.classAbiHash.size }
        return HashMap<String, ByteArray>(size).also { hashes ->
            filesToLoad.forEach {
                hashes.putAll(dataForFiles[it]!!.classAbiHash)
            }
        }
}
```

Reply:
Agree, it is more readable with your comment.